### PR TITLE
[CALCITE-3923] Refactor how planner rules are parameterized

### PR DESCRIPTION
**Pull request by julianhyde**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/RelOptRules.java [hunk 1]
```diff
@@ -157,18 +157,18 @@ private RelOptRules() {
...
+      ProjectToWindowRule.ProjectToLogicalProjectAndWindowRule.INSTANCE,
       FilterMergeRule.INSTANCE,
       DateRangeRules.FILTER_INSTANCE,
       IntersectToDistinctRule.INSTANCE);
 
   static final List<RelOptRule> ABSTRACT_RELATIONAL_RULES = ImmutableList.of(
-      FilterJoinRule.FILTER_ON_JOIN,
-      FilterJoinRule.JOIN,
+      FilterJoinRule.FILTER_ON_JOIN.get(),
+      FilterJoinRule.JOIN.get(),
```
Comment (rkondakov) [inline core/src/main/java/org/apache/calcite/plan/RelOptRules.java:167]: I am not sure that using suppliers is the most elegant way to get rid of deadlocks on class loading. As I understand the deadlock mechanics right, It is something extremely rare. What if we just deprecate using static fields 
```
FilterJoinRule.FILTER_ON_JOIN,
FilterJoinRule.JOIN
```
and force using static constants declared within the actual class (not a super class, i.e. FilterJoinRule) : 
```
FilterIntoJoinRule.INSTANCE,
JoinConditionPushRule.INSTANCE
```
What do you think?

Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/plan/RelOptRules.java]: I have made the change you suggested. We'll see whether anyone runs into deadlocks.

Fixed in 497fa05.

File: core/src/main/java/org/apache/calcite/plan/RelOptRules.java [hunk 2]
```diff
@@ -179,11 +179,9 @@ private RelOptRules() {
       SortRemoveRule.INSTANCE);
 
   static final List<RelOptRule> CONSTANT_REDUCTION_RULES = ImmutableList.of(
-      ReduceExpressionsRule.PROJECT_INSTANCE,
-      ReduceExpressionsRule.FILTER_INSTANCE,
-      ReduceExpressionsRule.CALC_INSTANCE,
-      ReduceExpressionsRule.WINDOW_INSTANCE,
-      ReduceExpressionsRule.JOIN_INSTANCE,
+      ReduceExpressionsRule.PROJECT_INSTANCE.get(),
+      ReduceExpressionsRule.FILTER_INSTANCE.get(), ReduceExpressionsRule.CALC_INSTANCE.get(),
```
[the remark this answers is not in the document]
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/plan/RelOptRules.java]: Fixed in 497fa05.

**Comment by julianhyde** [comment]
I have force-pushed the branch. The branch now contains several commits that should be merged before 1.24, and several that should be merged after 1.24 and before 1.25.

**Pull request closed without merging**